### PR TITLE
Updating memory optimization overview

### PR DESCRIPTION
**Pull request by SalmanMohammadi**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [x] update tests and/or documentation
- [ ] other (please add here)

Updating the memory optimization tutorial with `CPUOffloadingOptimizer`, 

```haskell
                                      .,'
                                   .'`.'
                                  .' .'
                      _.ood0Pp._ ,'  `.~ .q?00doo._
                  .od00Pd0000Pdb._. . _:db?000b?000bo.
                .?000Pd0000PP?000PdbMb?000P??000b?0000b.
              .d0000Pd0000P'  `?0Pd000b?0'  `?000b?0000b.
             .d0000Pd0000?'     `?d000b?'     `?00b?0000b.
             d00000Pd0000Pd0000Pd00000b?00000b?0000b?0000b
             ?00000b?0000b?0000b?b    dd00000Pd0000Pd0000P
             `?0000b?0000b?0000b?0b  dPd00000Pd0000Pd000P'
              `?0000b?0000b?0000b?0bd0Pd0000Pd0000Pd000P'
                `?000b?00bo.   `?P'  `?P'   .od0Pd000P'
                  `~?00b?000bo._  .db.  _.od000Pd0P~'
                      `~?0b?0b?000b?0Pd0Pd000PdP~'

```

### CODE REVIEW

File: docs/source/tutorials/memory_optimizations.rst [hunk 1]
```diff
@@ -16,9 +16,11 @@ To make things easy, we've summarized these components in the following table:
 
    ":ref:`glossary_precision`", "You'll usually want to leave this as its default ``bfloat16``. If you're struggling with training stability or accuracy due to precision, fp32 may help, but will significantly increase memory usage and decrease training speed."
    ":ref:`glossary_act_ckpt`", "Use when you're memory constrained and need to handle larger batch sizes or longer context lengths. Be aware that it may slow down training speed."
+   ":ref:`glossary_act_off`", "Similar to activation checkpointing, this can be used when memory constrained, but comes at the cost of training speed due to the overhead from moving tensors between GPU VRAM and CPU. This can also be used alongside activation checkpointing."
```
Comment (joecummings) [inline docs/source/tutorials/memory_optimizations.rst:19]: NITTTTT

```suggestion
   ":ref:`glossary_act_off`", "Similar to activation checkpointing, this can be used when memory constrained, but comes at the cost of training speed due to the overhead of moving tensors between GPU VRAM and CPU. This can also be used alongside activation checkpointing."
```

File: docs/source/tutorials/memory_optimizations.rst [hunk 2]
```diff
@@ -16,9 +16,11 @@ To make things easy, we've summarized these components in the following table:
...
    ":ref:`glossary_grad_accm`", "Helpful when memory-constrained to simulate larger batch sizes. Often preferable to activation checkpointing for better training speed."
    ":ref:`glossary_low_precision_opt`", "When you need to further reduce memory usage beyond using ``bf16`` by reducing the precision in the optimizer states. Note that lower precision optimizers may reduce training stability/accuracy."
    ":ref:`glossary_opt_in_bwd`", "Helps reduce memory usage when using stateful optimizers, particularly when full-finetuning large models with high gradient memory usage. This is not compatible with ``gradient_accumulation_steps``, so training may slow down due to reduced model throughput."
+   ":ref:`glossary_cpu_offload`", "Offloads optimizer states, and, optionally, gradients, to CPU, and performs optimizer steps on CPU. This can be used to significantly reduce GPU memory usage at the cost of CPU RAM and training speed, as CPU optimizer steps can be slow and bottleneck training performance."
```
Comment (joecummings) [inline docs/source/tutorials/memory_optimizations.rst:23]: NITTTTTT

```suggestion
   ":ref:`glossary_cpu_offload`", "Offloads optimizer states and (optionally) gradients to CPU, and performs optimizer steps on CPU. This can be used to significantly reduce GPU memory usage at the cost of CPU RAM and training speed, as CPU optimizer steps can be slow and bottleneck training performance."
```

File: docs/source/tutorials/memory_optimizations.rst [hunk 3]
```diff
@@ -208,10 +213,56 @@ To understand how this works, we encourage you to read through the relevant PyTo
...
+We've mentioned above the concept of optimizer states - memory used by the stateful optimizers to maintain a state of gradient statistics, and
+model gradients - tensors used to store gradients when we perform model backwards passes. We support using CPU offloading in our single-device recipes
+through the `CPUOffloadOptimizer <https://github.com/pytorch/ao/tree/main/torchao/prototype/low_bit_optim#optimizer-cpu-offload>`_ from ``torchao``.
+
+This optimizer can wrap any base optimizer and works by keeping the optimizer states and performing the optimizer step on CPU, thus reducing
+GPU memory usage by the size of the optimizer states. Additionally, we can also offload gradients to the CPU by using `offload_gradients=True`.
+
+*Sounds great! How do I use it?*
+
+To use this optimizer in your recipes, set the ``optimizer`` key in your config to :class:`torchao.prototype.low_bit_optim.CPUOffloadOptimizer`, which
```
Comment (joecummings) [inline docs/source/tutorials/memory_optimizations.rst:236]: Can we include a Python code block here too?

File: docs/source/tutorials/memory_optimizations.rst [hunk 4]
```diff
@@ -208,10 +213,56 @@ To understand how this works, we encourage you to read through the relevant PyTo
...
+    _component_: torchao.prototype.low_bit_optim.CPUOffloadOptimizer
+    offload_gradients: True
+    # additional key-word arguments can be passed to torch.optim.AdamW
+    lr: 4e-5
+
+Some helpful hints from the ``torchao`` `CPUOffloadOptimizer page <https://github.com/pytorch/ao/tree/main/torchao/prototype/low_bit_optim#optimizer-cpu-offload>`_:
+
+* The CPU optimizer step is often the bottleneck when optimizer CPU offload is used. To minimize the slowdown, it is recommended to (1) use full ``bf16`` training so that parameters, gradients, and optimizer states are in ``bf16``; and (2) give GPU more work per optimizer step (e.g. larger batch size with activation checkpointing, gradient accumulation).
+* Gradient accumulation should always be set to 1 when ``offload_gradients=True``, as gradients are cleared on GPU every backward pass.
+* This optimizer works by keeping a copy of parameters and pre-allocating gradient memory on CPU. Therefore, expect your RAM usage to increase by 4x model size.
```
Comment (joecummings) [inline docs/source/tutorials/memory_optimizations.rst:262]: Is it exactly 4x model size? Or roughly?

**Review by joecummings — APPROVED** [review]
I am the King of All Nits, bow before me

File: docs/source/tutorials/memory_optimizations.rst (hunk 4 above)
Comment (SalmanMohammadi) [reply docs/source/tutorials/memory_optimizations.rst:262]: According to the ao docs yeah 

> To minimize the amount of CPU<->GPU data transfer, we keep a copy of parameters and pre-allocate gradients memory on CPU. Therefore, expect your RAM usage to increase by 2x model size + optimizer state (which is 2x model size for Adam).

and since we always use adam it is 4x
